File: JULY-2025/27-07-2025/p2210.rb
# @param {Integer[]} nums
# @return {Integer}
def count_hill_valley(nums)
    hills, valley = 0, 0
    nums.each_with_index do |num, index|
        # Skip first and last as they are neither hill or valley
        if index == 0 || index == nums.length - 1
            next
        end

        # Go to last iteration of repeating character.
        if index + 1 < nums.length && nums[index] == nums[index + 1]
            next
        end

        # Go left to check for valley or hill
        left, right = index - 1, index + 1
        
        while left > 0 && nums[index] == nums[left]
            left -= 1
        end

        # Check for valley and hills
        valley = valley + 1 if nums[index] < nums[left] && nums[index] < nums[right]
        hills = hills + 1 if nums[index] > nums[left] && nums[index] > nums[right]
    end
    return valley + hills
end

File: JULY-2025/27-07-2025/p2210.php
class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer
     */
    function getLeft($nums, $i){
        if(isset($nums[$i - 1])){
            if($nums[$i] == $nums[$i - 1])
                return $this->getLeft($nums, --$i);
            else
                return $i - 1;
        }
        else{
            return $i;
        }
    }

    function countHillValley($nums) {
        $countNums = count($nums) - 1;
        $valleys = 0;
        $hills = 0;
        for($i = 1; $i < $countNums; $i++){
                $left = $this->getLeft($nums, $i);

                $valleys += $nums[$i] < $nums[$left] && $nums[$i] < $nums[$i + 1] ? 1 : 0;
                $hills += $nums[$i] > $nums[$left] && $nums[$i] > $nums[$i + 1] ? 1 : 0;
        }

        return $valleys + $hills;
    }
}